src/core/storage/fileio/s3_api.cpp

```diff
@@ -513,13 +514,31 @@ list_objects_response list_objects_impl(s3url parsed_url, std::string proxy,
 
       } else {
         auto error = outcome.GetError();
-        // Unlike CoreErrors, S3Error Never retries. Use Http code instead.
-        // check aws-cpp-sdk-s3/source/S3Error.cpp
-        if (error.GetResponseCode() ==
-            Aws::Http::HttpResponseCode::TOO_MANY_REQUESTS) {
-          n_retry++;
+        /*
+         * Unlike CoreErrors, S3Error Never retries on S3 errors.
+         * Retry can be based on HTTP code or HTTP body.
+         *
+         * 1. if SDK uses HTTP code, e.g., ShouldRetry return true on 429.
+         * We don't need to retry on our own since SDK already made decision to
+         * retry based on HTTP code.
+         *
+         * 2. if SDK doesn't use HTTP code but HTTP body, we check HTTP on our
+         * own if SDK doesn't think it should retry based on messages in HTTP
+         * body. Check https://guihao-liang.github.io/2020-04-12-aws-s3-retry/
+         *
+         * */
+        if (!error.ShouldRetry()) {
+          // SDK didn't retry for us, check retry on our own decisions
+          if (error.GetResponseCode() ==
+              Aws::Http::HttpResponseCode::TOO_MANY_REQUESTS) {
+            n_retry++;
+          } else {
+            // stop retry immediately
+            n_retry = 3;
+          }
 
-          if (n_retry == 3) {
+          // should be ==, use >= to suppress any surprises
```

Comment:
just want to let people know the why I use >=. Theoriactially, its max value should be 3.

Reply:
`to suppress any surprises` is kind of a strange phase. I don't think this require a comment but if you really want to add one I would say `// Defensively use >= rather than ==`. It probably make more sense as an inline comment.

Reply:
I think Toby's comment is a good one here, or just drop it -- it's obvious in this what's going on.